symbolic-debuginfo/src/sourcebundle.rs

```diff
@@ -389,6 +316,128 @@ impl SourceFileInfo {
+    pub fn ty(&self) -> SourceFileType {
+        self.file_info
+            .and_then(|x| x.ty())
+            .unwrap_or(SourceFileType::Source)
+    }
+
+    /// The contents of the source file as string, if it's available.
+    ///
+    /// Portable PDBs for instance will often have source information, but rely on
+    /// remote file fetching via Sourcelink to get to the contents.  In that case
+    /// a file descriptor is created, but the contents are missing and instead the
+    /// [`url`](Self::url) can be used.
+    pub fn contents(&self) -> Option<&str> {
+        self.contents.as_deref()
+    }
+
+    /// If available returns the URL of this source.
+    ///
+    /// For certain files this is the canoncial URL of where the file is placed.  This
+    /// for instance is the case for minified JavaScript files or source maps which might
+    /// have a canonical URL.  In case of portable PDBs this is also where you would fetch
+    /// the source code from if source links are used.
+    pub fn url(&self) -> Option<&str> {
+        if let Some(ref url) = self.remote_url {
+            Some(url)
+        } else {
+            self.file_info.and_then(|x| x.url())
+        }
+    }
+
+    /// If available returns the file path of this source.
+    ///
+    /// For source bundles that are a companion file to a debug file, this is the canonical
+    /// path of the source file.
+    pub fn path(&self) -> Option<&str> {
+        self.file_info.and_then(|x| x.path())
+    }
+
+    /// The debug ID of the file if available.
+    ///
+    /// For source maps or minified source files symbolic supports embedded debug IDs.  If they
+    /// are in use, the debug ID is returned from here.
+    pub fn debug_id(&self) -> Option<DebugId> {
+        self.file_info.and_then(|x| x.debug_id())
+    }
+
+    /// The source mapping URL reference of the file.
+    ///
+    /// This is used to refer to a source map from a minified file.  Only minified source files
+    /// will have a relationship to a source map.
+    pub fn source_mapping_url(&self) -> Option<&str> {
+        if let Some(file_info) = self.file_info {
+            if let Some(url) = file_info.source_mapping_url() {
+                return Some(url);
+            }
+        }
+        if let Some(ref contents) = self.contents {
+            if let Some(url) = discover_sourcemaps_location(contents) {
+                return Some(url);
+            }
+        }
+        None
+    }
+}
+
+/// Parses a sourceMappingURL comment in a file to discover a sourcemap reference.
+fn discover_sourcemaps_location(contents: &str) -> Option<&str> {
+    for line in contents.lines().rev() {
+        if line.starts_with("//# sourceMappingURL=") || line.starts_with("//@ sourceMappingURL=") {
+            let possible_sourcemap = line[21..].trim();
+            if possible_sourcemap.starts_with("data:application/json") {
+                return None;
+            }
```

Comment:
not sure if we actually want to return the `data` url here? Otherwise we would have to do that same detection logic in symbolicator again. CC @loewenheim wdyt, as you implemented the base64 encoded data urls?

Reply:
I think we should expose the data url here, yes.

Reply:
I'm not sure though if we should leave the data url prefix alone; if the URL is returned as a plain string, we will need to be able to recognize it later.